flamingo/scripts/hot_gas_plots.py

```diff
@@ -0,0 +1,460 @@
+xi_int = CubicSpline(z_tab, xi, bc_type=((2, 0.0), (2, 0.0)), extrapolate=False)
+
+ne_n = xe_int(Zmet) / (xe_int(Zmet) + xi_int(Zmet))
+
+@njit
+def compute_Xs_mu(elements_array):
+
+    all_elements = ['hydrogen', 'helium', 'carbon', 'nitrogen', 'oxygen', 'neon', 'magnesium', 'silicon', 'sulphur', 'calcium', 'iron']
+    element_masses = [1.00794, 4.002602, 12.0107, 14.0067, 15.9994, 20.1797, 24.3050, 28.0855, 32.065, 40.078, 55.845]
+    agtomic_num = [1.0, 2.0, 6.0, 7.0, 8.0, 10.0, 12.0, 14.0, 16.0, 20.0, 26.0]
+
+    # Hydrogen
+    ne_nH = np.full_like(elements_array[:, 0], 1.0)
+    ni_nH = np.full_like(elements_array[:, 0], 1.0)
+    mu = np.full_like(elements_array[:, 0], 0.5)
+
+    # Every other element
+    for i in range(len(all_elements) - 1):
+        ne_nH += elements_array[:, i + 1] * (element_masses[0] / element_masses[i + 1]) * (agtomic_num[i + 1] / agtomic_num[0])
+        ni_nH += elements_array[:, i + 1] * (element_masses[0] / element_masses[i + 1]) 
+        mu += elements_array[:, i + 1] / (agtomic_num[0] + agtomic_num[i + 1])
+
+    return ne_nH, ni_nH, mu
+
+def read_hot_gas_props(run_directory, snapshot_name, cat_name, obs_data_dir, chunk):
+    
+    elements = ['hydrogen', 'helium', 'carbon', 'nitrogen', 'oxygen', 'neon', 'magnesium', 'silicon', 'iron']
+    all_elements = ['hydrogen', 'helium', 'carbon', 'nitrogen', 'oxygen', 'neon', 'magnesium', 'silicon', 'sulphur', 'calcium', 'iron']
+    SulphurOverSilicon = 0.6054160
+    CalciumOverSilicon = 0.0941736
+
+    catalogue = load_catalogue(run_directory + '/' + cat_name)
+    groups = load_groups(run_directory + '/' + cat_name.replace('properties', 'catalog_groups'), catalogue=catalogue)
+
+    df_chunk = pd.DataFrame()
+    halo_n = 1
+    tot_h = len(chunk.index)
+    for index, row in chunk.iterrows():
+        print('TaskID: ', multiprocessing.current_process()._identity[0], 'Processing halo = ', int(row.GroupNumber), '(', halo_n, '/', tot_h, ')')
+        particles, unbound_particles = groups.extract_halo(halo_id=int(row.GroupNumber))
+        data, bmask = to_swiftsimio_dataset(particles, run_directory + '/' + snapshot_name, generate_extra_mask=True)
+        udata, umask = to_swiftsimio_dataset(unbound_particles, run_directory + '/' + snapshot_name, generate_extra_mask=True)
+        mask = np.logical_or(bmask.gas, umask.gas)
+
+        meta = data.metadata        
+        BoxSize = meta.boxsize[0].value
+        cosmo =  meta.cosmology
+        aexp = meta.a
+        z = meta.redshift
+        ez = cosmo.efunc(z)
+
+        df_elements = pd.DataFrame()
+        for element in elements:
+            df_elements[element] = getattr(data.gas.smoothed_element_mass_fractions, element)
+
+        df_gas = pd.DataFrame(data.gas.coordinates.value, columns=['x', 'y', 'z'])
+        df_gas['temp'] = data.gas.temperatures.value
+        df_gas['mass'] = data.gas.masses.to('Msun').value.astype(np.float64) * u.Msun.to(u.gram)
```

Comment:
Is there a specific reason why you are not simply using `data.gas.masses.to("gram").value`? Is this a rounding issue?

Reply:
Yes, the problem is that the values are really small in grams, and rounding issues arise further down in the calculations and would return either inf of nan. 

I tried different hacks here, but if I remember correctly, the only way I made this work is to convert to Msun first using float64, the convert to grams. But there might be a better solution.